### ISSUE TITLE
Transaction with LoadWith

### ISSUE THREAD

**Issue by dzyranov** [linq2db/linq2db#482]
Load phantom record
linq2Db 1.0.7.4, npgsql 3.1.9.0, postgres 9.6
```
db.BeginTransaction(System.Data.IsolationLevel.Serializable);

// delete all
db.GetTable<CUserAchievement>().Delete();

//  ok  ss.Count = 0
var ss = db.GetTable<CUserAchievement>().ToList();

//  phantom record  a.Count != 0
a = db.GetTable<CUser>().LoadWith(p => p.Achievements).Where(p => p.Caption == "test").Single().Achievements;
```

**Comment by sdanyliv** [comment]
And without transaction?
If without transaction result is correct - it can be postgres bug.
Try to reproduce in query analyser

**Comment by dzyranov** [comment]
Inside transaction

**Comment by sdanyliv** [comment]
I think linq2db do not have special code for handling records inside transactions. It can be postgres bug.

**Comment by dzyranov** [comment]
var ss = db.GetTable<CUserAchievement>().ToList();
generate select query

a = db.GetTable<CUser>().LoadWith(p => p.Achievements).Where(p => p.Caption == "test").Single().Achievements;
generate anonymnos block

LoadWith is multi thread ?

**Comment by dzyranov** [comment]
I think
var user = db.GetTable().LoadWith(p => p.Achievements).Where(p => p.Caption == "test").Single();
should be equivalence
var user = db.GetTable<CUser>().Single(p => p.Caption == "test");
user.Achievements = db.GetTable<CUserAchievement>().Where(p => p.UserId == p.Id).ToList();
but in practice it is not

**Comment by jogibear9988** [comment]
why should it? i think  the first should result in a query with a join, the second in two queries

**Comment by dzyranov** [comment]
No, LoadWith generate two query, 1 query in transaction, 2 not in transaction

**Comment by jogibear9988** [comment]
no, i'm wrong, you only call tolist after the second

**Comment by sdanyliv** [comment]
@dzyranov, for now it is more clear. Looks like we can not open another record reader until we not closed previous, so it means new connection. Needs deeper look how to share transaction between progres connections, if it is possible.

**Comment by ili** [comment]
> LoadWith generate two query, 1 query in transaction, 2 not in transaction

this is bug, and should be fixed

**Comment by ili** [comment]
This is by design now. 

To use it in this way you need to enable [MARS](https://msdn.microsoft.com/en-us/library/h32h3abf(v=vs.110).aspx)

Now this is supported only for MS Sql Server. I'll look forward to check if we can support MARS for other providers.

**Comment by i-sinister** [comment]
This is very strange 'design' which effectively renders associations unusable since it brakes consistency and isolation (if apply these to read operations). I've encountered this issue when trying to use associations with in-memory SQLite and it simply does not work become with 'in-memory' DB opening new connection create new empty DB. But I consider my self lucky, for this was much easier to reproduce and analyze then phantom reads I'd have in production.

Problem is AssociatedTableContext.SubQueryHelper<T>.ExecuteSubQuery is cloning DataConnection, which in turn calls Clone on the ADO.NET connection and results in associations being loaded in different connection (hence transaction).

Proper (and I guess the only possible) to handle 1:M associations would be to cache results of the initial query first, open another IDataReader to load associations and then construct objects using cached results of the first query and results from the reader. This, however, will brake the niceness of kinda lazy current IEnumerable implementation

**Comment by nzandy** [comment]
Bump. I have just been migrating our code from SimpleData to Linq2Db (yay!) however now our .LoadWith() commands inside TransactionScope are failing. Are there any alternatives other than enabling MARS application wide?

**Comment by sdanyliv** [comment]
@nzandy, please post simple model and reproducing linq query. Thanks!
Sorry for delays but we are on vacations for two weeks. Please not that LoadWith has limited support and we have plans to improve that.

**Comment by nzandy** [comment]
`		public List<SegmentTemplate> FindSegmentTemplates(int providerId, SegmentType segmentType)
		{
			using (var db = new PathwayDataConnection())
			{
				if (segmentType == SegmentType.ClinicLinked)
				{
					return db.SegmentTemplate
						.LoadWith(template => template.SegmentTemplateSessionCodes)
						.Where(template => template.ProviderId == providerId && template.Type == segmentType).ToList();
				}
				return db.SegmentTemplate.Where(template => template.ProviderId == providerId && template.Type == segmentType).ToList();
			}
		}`

And the calling code that wraps with transaction:

`	using (var transaction = new TransactionScope())
			{
				result = _pathwaySegmentTemplateRepository.FindSegmentTemplates(OtherProviderId, segmentType).ToList();
				transaction.Complete();
			}`

resulted in the error.

**Comment by MaceWindu** [comment]
Fixed with new eager load implementation #1756
